### ISSUE TITLE
OnDemand VHDS makes the first request returns 404.

### ISSUE THREAD

**Issue by WeavingGao** [envoyproxy/envoy#17891]
*Description*:
> I used OnDemand VHDS to dynamically obtain the virtual host configuration. When a  request to a certain host comes for the first time, the request of the `route` is `nullptr`, then the `on demand filter` will send a VHDS request, and execute its callback `onRouteConfigUpdateCompletion` when the VHDS response is received. If the request to the host doesn't have a body, it works fine because a new stream will be created and the request's route will be re-calculated. However if the request has a body,   it won't re-create a stream but just continue to next filter, thus the request's `route` is still `nullptr` since the `snapped_route_config_` is still the old one. Finally the `router` filter will send a local response of 404 to downstream and set the response flag to `NR`.

cc @dmitri-d @htuch @lambdai

**Comment by alyssawilk** [comment]
Yeah, the router now supports internal redirects with a fully read body (https://github.com/envoyproxy/envoy/pull/15634) but the on demand filter does not yet use that new code (https://github.com/envoyproxy/envoy/blob/main/source/extensions/filters/http/on_demand/on_demand_update.cc#L63)

stopalliteration + explicit read disable might work for HTTP/1.1 downstream but due to how HTTP/2 buffering works, it will still end up with a body buffer and still trip checks in the on demand filter.  I think adding support for fully read bodies (as the router filter now has) would be a cleaner next step.

**Comment by uglycow** [comment]
> On-demand/vhds supports requests without a body only at the moment.
> 
> It should be possible to do an internal redirect with a body now, provided `envoy.reloadable_features.internal_redirects_with_body` has been enabled and the request has been fully read (here: https://github.com/envoyproxy/envoy/blob/main/source/common/http/conn_manager_impl.cc#L1684)? In this case RouteConfig update will need to be initiated from `OnDemandRouteUpdate::decodeData()` when end_stream is received. @alyssawilk: did I get this right?

is this fixed now ?

**Comment by maoyutao** [comment]
is this fixed now ?

**Comment by sichan-vonage** [comment]
How's this going @alyssawilk any progress made on this?

**Comment by alyssawilk** [comment]
This is unassigned - no one has picked it up thusfar sorry

**Comment by hochuenw-dd** [comment]
Hey folks, since this bug has remained unresolved for over three years, I'm curious - how are you working around it to dynamically fetch virtual hosts? cc @alyssawilk @sichan-vonage @maoyutao @uglycow @WeavingGao @dmitri-d
If there is no workaround, does this mean there is still no solid approach to NOT require users to predefine their egress dependencies of their service?

**Comment by steveohara** [comment]
Getting this problem fixed is very important to us, which means that we potentially we have funding available if someone with the right skill set wants to come forward with a proposal??
@alyssawilk @sichan-vonage @maoyutao @uglycow @WeavingGao @dmitri-d
